// md-accordion/accordionpanel.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var MdAccordion = (function () {
    function MdAccordion() {
        this.mdClass = '';
        this.close = new core_1.EventEmitter();
        this.open = new core_1.EventEmitter();
        this.tabs = [];
    }
    /**
     * Add or append tab in accordion
     * @param tab object of MdAccordionTab
     */
    MdAccordion.prototype.addTab = function (tab) {
        this.tabs.push(tab);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], MdAccordion.prototype, "multiple", void 0);
    __decorate([
        core_1.Input('class'), 
        __metadata('design:type', String)
    ], MdAccordion.prototype, "mdClass", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], MdAccordion.prototype, "close", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], MdAccordion.prototype, "open", void 0);
    MdAccordion = __decorate([
        core_1.Component({
            // moduleId: module.id,
            selector: 'md-accordion',
            template: "<ng-content></ng-content>",
            host: {
                '[class]': 'mdClass',
                '[class.md-accordion]': 'true'
            },
            styles: ["\n    .md-accordion { display: block; }\n  "],
            encapsulation: core_1.ViewEncapsulation.None
        }), 
        __metadata('design:paramtypes', [])
    ], MdAccordion);
    return MdAccordion;
}());
exports.MdAccordion = MdAccordion;
