// md-accordion/accordion.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var accordionpanel_1 = require('./accordionpanel');
var accordiontab_1 = require('./accordiontab');
var accordionpanel_2 = require('./accordionpanel');
exports.MdAccordion = accordionpanel_2.MdAccordion;
var accordiontab_2 = require('./accordiontab');
exports.MdAccordionTab = accordiontab_2.MdAccordionTab;
exports.ACCORDION_DIRECTIVES = [accordionpanel_1.MdAccordion, accordiontab_1.MdAccordionTab];
var MdAccordionModule = (function () {
    function MdAccordionModule() {
    }
    MdAccordionModule = __decorate([
        core_1.NgModule({
            imports: [common_1.CommonModule],
            exports: exports.ACCORDION_DIRECTIVES,
            declarations: exports.ACCORDION_DIRECTIVES,
        }), 
        __metadata('design:paramtypes', [])
    ], MdAccordionModule);
    return MdAccordionModule;
}());
exports.MdAccordionModule = MdAccordionModule;
