// angular-material-datepicker/src/datepicker/lang-de.ts
export const LANG_DE = {
    weekDays: [
        {
            letter: "S",
            short: "So.",
            full: "Sonntag"
        },
        {
            letter: "M",
            short: "Mo.",
            full: "Montag"
        },
        {
            letter: "D",
            short: "Di.",
            full: "Dienstag"
        },
        {
            letter: "M",
            short: "Mi.",
            full: "Mittwoch"
        },
        {
            letter: "D",
            short: "Do.",
            full: "Donnerstag"
        },
        {
            letter: "F",
            short: "Fr.",
            full: "Freitag"
        },
        {
            letter: "S",
            short: "Sa.",
            full: "Samstag"
        }
    ],
    months: [
        {
            short: "Jan.",
            full: "Januar"
        },
        {
            short: "Feb.",
            full: "Februar"
        },
        {
            short: "März",
            full: "März"
        },
        {
            short: "Apr.",
            full: "April"
        },
        {
            short: "Mai",
            full: "Mai"
        },
        {
            short: "Juni",
            full: "Juni"
        },
        {
            short: "Juli",
            full: "Juli"
        },
        {
            short: "Aug.",
            full: "August"
        },
        {
            short: "Sep.",
            full: "September"
        },
        {
            short: "Okt.",
            full: "Oktober"
        },
        {
            short: "Nov.",
            full: "November"
        },
        {
            short: "Dez.",
            full: "Dezember"
        }
    ]
}
