// angular-material-datepicker/src/datepicker/lang-en.ts
export const LANG_EN = {
    weekDays: [
        {
            letter: "S",
            short: "Sun",
            full: "Sunday"
        },
        {
            letter: "M",
            short: "Mon",
            full: "Monday"
        },
        {
            letter: "T",
            short: "Tue",
            full: "Tuesday"
        },
        {
            letter: "W",
            short: "Wed",
            full: "Wednesday"
        },
        {
            letter: "T",
            short: "Thu",
            full: "Thursday"
        },
        {
            letter: "F",
            short: "Fri",
            full: "Friday"
        },
        {
            letter: "S",
            short: "Sat",
            full: "Saturday"
        }
    ],
    months: [
        {
            short: "Jan",
            full: "January"
        },
        {
            short: "Feb",
            full: "February"
        },
        {
            short: "Mar",
            full: "March"
        },
        {
            short: "Apr",
            full: "April"
        },
        {
            short: "May",
            full: "May"
        },
        {
            short: "Jun",
            full: "June"
        },
        {
            short: "Jul",
            full: "July"
        },
        {
            short: "Aug",
            full: "August"
        },
        {
            short: "Sep",
            full: "September"
        },
        {
            short: "Oct",
            full: "October"
        },
        {
            short: "Nov",
            full: "November"
        },
        {
            short: "Dec",
            full: "December"
        }
    ]
}
